p2p/net/swarm/swarm.go

```diff
@@ -541,18 +541,10 @@ func (s *Swarm) Backoff() *DialBackoff {
 
 // notifyAll sends a signal to all Notifiees
 func (s *Swarm) notifyAll(notify func(network.Notifiee)) {
-	var wg sync.WaitGroup
-
 	s.notifs.RLock()
-	wg.Add(len(s.notifs.m))
 	for f := range s.notifs.m {
-		go func(f network.Notifiee) {
-			defer wg.Done()
-			notify(f)
```

Comment:
I think this old code is actually a bit broken wrt to ordering of events.

Since it only acquires a read lock, there's nothing stopping the notifiee from getting an event out of order since the notify call happens in async in a goroutine.

I think we should use a normal mutex here instead of RWMutex to guarantee this ordering.

Reply:
That lock is protecting the notifiee list (and is global). The lock you're looking for is https://github.com/libp2p/go-libp2p/blob/c7928312d6c90635c835ac3e4bd5534b1763189a/p2p/net/swarm/swarm.go#L303-L304.

Reply:
Ah okay, thank you!